bpf/lib/icmp6.h

```diff
@@ -534,6 +534,10 @@ bool icmp6_ndisc_validate(struct __ctx_buff *ctx, const struct ipv6hdr *ip6,
 	struct ethhdr *eth = ctx_data(ctx);
 	union macaddr *dmac;
 
+	/* On l2-less devices there is no MAC address, and we cannot proceed */
+	if (sizeof(struct ethhdr) != ETH_HLEN)
+		return false;
+
```

Comment:
Looks good as quick fix! But I think we should be able to lift this even further up, and just skip `handle_l2_announcement()` entirely on a L3 device? Attempting to advertise `CONFIG(interface_mac)` makes no sense, even for IPv4 ...

Reply:
I looked into a bit more, and its just a difficult and confusing situation. To summarize:
* In bpf_host.c its dynamic 0 or `__ETH_HLEN` (14), depending on the device its attached to
* In bpf_wireguard.c its static 0, because wg devices are l2-less
* In all others its static `__ETH_HLEN` (14)

For thus for library functions it depends on which .c file is the current root. And you as developer have to have knowledge of who uses the logic to know if your macro is static or dynamic.

We use `ETH_HLEN` about 150 times, so I will drop the idea of `curDevL3` I don't think it will help.

As for these changes, perhaps comparing `ETH_HLEN` to `__ETH_HLEN` or 0 is more sound than comparing to `sizeof(struct ethhdr)` even though its technically the same

Reply:
> As for these changes, perhaps comparing `ETH_HLEN` to `__ETH_HLEN` or 0 is more sound than comparing to `sizeof(struct ethhdr)` even though its technically the same

ack, think I'll send a tree-wide cleanup to introduce a `is_l3_device()` helper (which checks for `ETH_HLEN == 0`). But let's merge this as-is, no need to over-complicate a fix which needs backporting.

Reply:
Playing around in https://github.com/cilium/cilium/pull/41750.